examples/vision/metric_learning.py

```diff
@@ -22,290 +24,380 @@
-from tensorflow import keras
 from tensorflow.keras import layers
+from tensorflow.keras.optimizers import Adam
 
-"""
-## Dataset
-
-For this example we will be using the
-[CIFAR-10](https://www.cs.toronto.edu/~kriz/cifar.html) dataset.
-"""
+import tensorflow_similarity as tfsim
+from tensorflow_similarity.layers import MetricEmbedding  # row wise L2 norm
+from tensorflow_similarity.losses import MultiSimilarityLoss  # specialized similarity loss
+from tensorflow_similarity.models import SimilarityModel  # TF model with additional features
+from tensorflow_similarity.samplers import TFDatasetMultiShotMemorySampler  # Get and samples TF dataset catalog
+from tensorflow_similarity.visualization import confusion_matrix  # matching performance
+from tensorflow_similarity.visualization import projector  # allows to interactively explore the samples in 2D space
+from tensorflow_similarity.visualization import viz_neigbors_imgs  # neighbors visualization
 
-from tensorflow.keras.datasets import cifar10
 
-(x_train, y_train), (x_test, y_test) = cifar10.load_data()
+tfsim.utils.tf_cap_memory()
 
-x_train = x_train.astype("float32") / 255.0
-y_train = np.squeeze(y_train)
-x_test = x_test.astype("float32") / 255.0
-y_test = np.squeeze(y_test)
+print('TensorFlow:', tf.__version__)
+print('TensorFlow Similarity', tfsim.__version__)
 
 """
-To get a sense of the dataset we can visualise a grid of 25 random examples.
-
-
+# Dataset Samplers
+
+We will be using the
+[CIFAR-10](https://www.tensorflow.org/datasets/catalog/cifar10) dataset for this
+tutorial.
+
+For a similarity model to learn efficiently, each batch must contains at least 2
+examples of each class.
+
+To make this easy, tf_similarity offers `Samplers()` that enable you to set both
+the number of classes and the minimum number of examples of each class per
+batch.
+
+The train and test datasets will be created using the
+`TFDatasetMultiShotMemorySampler()`. This creates a sampler that loads datasets
+from [tensorflow datasets](https://www.tensorflow.org/datasets) and yields
+batches containing a target number of classes and a target number of examples
+per class. Additionally, we can restrict the sampler to only yield the subset of
+classes defined in class_list, enabling us to train on a subset of the classes
+and then test how the embedding generalizes to the unseen classes. This can be
+useful when working on few-shot learning problems.
+
+The following cell creates a train_ds sample that:
+
+* Loads the cifar10 dataset from tfds and then takes the
+examples_per_class_per_batch.
+* Normalizes the images by applying the preprocess function to each image.
+* Ensures the sampler restricts the classes to those defined in the class_list.
+* Ensures each batch contains 4 different classes with 16 examples each.
+
+We also create a train_ds in the same way, but we limit the total number of
+examples per class to 100 and the examples per class per batch is set to the
+default of 2.
```

Comment:
This is not what the code do below. Please make sure its consistent.

Reply:
This was meant to refer to the test_ds set. I'll add a similar section to describe the train_ds params.